graph/manifest.go

```diff
@@ -8,62 +8,106 @@ import (
 	"github.com/docker/distribution/digest"
 	"github.com/docker/docker/registry"
 	"github.com/docker/docker/trust"
-	"github.com/docker/docker/utils"
 	"github.com/docker/libtrust"
 )
 
-// loadManifest loads a manifest from a byte array and verifies its content.
-// The signature must be verified or an error is returned. If the manifest
-// contains no signatures by a trusted key for the name in the manifest, the
-// image is not considered verified. The parsed manifest object and a boolean
-// for whether the manifest is verified is returned.
-func (s *TagStore) loadManifest(manifestBytes []byte, dgst, ref string) (*registry.ManifestData, bool, error) {
+// loadManifest loads a manifest from a byte array and verifies its content,
+// returning the local digest, the manifest itself, whether or not it was
+// verified. If ref is a digest, rather than a tag, this will be treated as
+// the local digest. An error will be returned if the signature verification
+// fails, local digest verification fails and, if provided, the remote digest
+// verification fails. The boolean return will only be false without error on
+// the failure of signatures trust check.
+func (s *TagStore) loadManifest(manifestBytes []byte, ref string, remoteDigest digest.Digest) (digest.Digest, *registry.ManifestData, bool, error) {
 	sig, err := libtrust.ParsePrettySignature(manifestBytes, "signatures")
 	if err != nil {
-		return nil, false, fmt.Errorf("error parsing payload: %s", err)
+		return "", nil, false, fmt.Errorf("error parsing payload: %s", err)
 	}
 
 	keys, err := sig.Verify()
 	if err != nil {
-		return nil, false, fmt.Errorf("error verifying payload: %s", err)
+		return "", nil, false, fmt.Errorf("error verifying payload: %s", err)
 	}
 
 	payload, err := sig.Payload()
 	if err != nil {
-		return nil, false, fmt.Errorf("error retrieving payload: %s", err)
+		return "", nil, false, fmt.Errorf("error retrieving payload: %s", err)
 	}
 
-	var manifestDigest digest.Digest
+	// TODO(stevvooe): It would be a lot better here to build up a stack of
+	// verifiers, then push the bytes one time for signatures and digests, but
+	// the manifests are typically small, so this optimization is not worth
+	// hacking this code without further refactoring.
 
-	if dgst != "" {
-		manifestDigest, err = digest.ParseDigest(dgst)
+	var localDigest digest.Digest
+
+	// verify the local digest, if present in tag
+	if dgst, err := digest.ParseDigest(ref); err != nil {
```

Comment:
Also, random question: what happens if someone specifies a tag that matches the format of sha256:xxxxxxxxxxx?